lndclient/versioner_client.go

```diff
@@ -0,0 +1,179 @@
+package lndclient
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/lightningnetwork/lnd/lnrpc/verrpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	// ErrVersionCheckNotImplemented is the error that is returned if the
+	// version RPC is not implemented in lnd. This means the version of lnd
+	// is lower than v0.10.0-beta.
+	ErrVersionCheckNotImplemented = errors.New("version check not " +
+		"implemented")
+
+	// ErrVersionIncompatible is the error that is returned if the connected
+	// lnd instance is not supported.
+	ErrVersionIncompatible = errors.New("version incompatible")
+
+	// ErrBuildTagsMissing is the error that is returned if the
+	// connected lnd instance does not have all built tags activated that
+	// are required.
+	ErrBuildTagsMissing = errors.New("build tags missing")
+)
+
+// VersionerClient exposes the version of lnd.
+type VersionerClient interface {
+	// GetVersion returns the version and build information of the lnd
+	// daemon.
+	GetVersion(ctx context.Context) (*verrpc.Version, error)
+}
+
+type versionerClient struct {
+	client      verrpc.VersionerClient
+	readonlyMac serializedMacaroon
+}
+
+func newVersionerClient(conn *grpc.ClientConn,
+	readonlyMac serializedMacaroon) *versionerClient {
+
+	return &versionerClient{
+		client:      verrpc.NewVersionerClient(conn),
+		readonlyMac: readonlyMac,
+	}
+}
+
+// GetVersion returns the version and build information of the lnd
+// daemon.
+//
+// NOTE: This method is part of the VersionerClient interface.
+func (v *versionerClient) GetVersion(ctx context.Context) (*verrpc.Version,
+	error) {
+
+	rpcCtx, cancel := context.WithTimeout(
+		v.readonlyMac.WithMacaroonAuth(ctx), rpcTimeout,
+	)
+	defer cancel()
+	return v.client.GetVersion(rpcCtx, &verrpc.VersionRequest{})
+}
+
+// CheckVersionCompatibility makes sure the connected lnd node has the correct
+// version and required build tags enabled.
+//
+// NOTE: This check will **never** return a non-nil error for a version of
+// lnd < 0.10.0 because any version previous to 0.10.0 doesn't have the version
+// endpoint implemented!
+func CheckVersionCompatibility(client VersionerClient, major, minor,
```

Comment:
Does this need to be exported? Is there a use case to check the version again after checking it on startup? If not exported, it can probably be moved near the startup code.